### ISSUE TITLE
.NET 8 HttpContext is null when it is not in .NET 6

### ISSUE THREAD

**Issue by MarkMilfordMe** [Azure/azure-functions-dotnet-worker#2372]
### Description

As per problem here: https://github.com/dotnet/aspnetcore/issues/53143 

### Steps to reproduce

Project repo with .NET 8 and .NET 6 demonstration here: 
https://github.com/MarkMilfordMe/AzureFunctionsHttpContextorIssue

If you run the .NET 8 project, with url: /api/Function1?ValueFromParameter=test if fails on Program.cs line 22
.NET 6 project runs fine, with url: /api/Function1?ValueFromParameter=test
I've tried to make the code as identical as possible

Issue is in .NET 8 it doesn't make IHttpContextAccessor.HttpContext available during dependency injection.

**Comment by rhythmnewt** [comment]
Came across same issue, after migrating from .net 6 to .net 8 the IHttpContextAccessor is null when retrieved from context.

![image](https://github.com/Azure/azure-functions-dotnet-worker/assets/1319322/be09a178-b32a-4c95-aff2-6002ca54ad4d)

**Comment by Tech-Sohail** [comment]
I am also facing the same issue when I am trying to upgrade my projects from Net6 to Net8.
I can initialize HttpContextAccessor by using services.AddHttpContextAccessor()
But then HttpContextAccessor.HttpContext is null.

Any idea when this issue is going to be fixed?

Note: We didn't need to initialize HttpContextAccessor in Net6. It is automatically getting initialized during each function endpoint call.

**Comment by cosminvladutu** [comment]
Same issue like @Tech-Sohail just described, here also :(

**Comment by MarkMilfordMe** [comment]
> Same issue like @Tech-Sohail just described, here also :(
> 
> As far as I understand, the problem is that I am not using an HTTPTrigger. If you are using an Event trigger or a Timer trigger, you won't have a context... IHttpContextAccessor and HttpContext are not available in these cases. It seems the idea is that if you have information in the context, that info should be in the event itself.

No, this issue is purely about the lack of HTTP Context on a http trigger, I would never expect to have a http trigger on event or Timer, in .NET 6 we uses Queue triggers extensively but there is no http context in those circumstances and that is expected behaviour.  

Still waiting on anything from MS on this, so far seems to just have been added and removed from various labels, I rechecked the latest nuget packages recently (see sample project) and its still ongoing to latest releases

**Comment by patrick-elmore** [comment]
@MarkMilfordMe I don't think that this is a .NET 8 vs .NET 6 issue.  It appears to be an in-process vs out-of-process function issue.  Your .NET 6 demo project is in-process and your .NET 8 demo project is out-of-process.  If you were to change your .NET 6 project to out-of-process, I suspect you would see the same behavior.

**Comment by MarkMilfordMe** [comment]
> @MarkMilfordMe I don't think that this is a .NET 8 vs .NET 6 issue. It appears to be an in-process vs out-of-process function issue. Your .NET 6 demo project is in-process and your .NET 8 demo project is out-of-process. If you were to change your .NET 6 project to out-of-process, I suspect you would see the same behavior. We are currently seeing the same thing when doing a straight in-process to out-of-process migration in a .NET 6 function application.

Thanks, I'll try that, guess still means we can't update to .NET 8 though, as .NET 8 doesn't support in-process to my knowledge.

**Comment by patrick-elmore** [comment]
> > @MarkMilfordMe I don't think that this is a .NET 8 vs .NET 6 issue. It appears to be an in-process vs out-of-process function issue. Your .NET 6 demo project is in-process and your .NET 8 demo project is out-of-process. If you were to change your .NET 6 project to out-of-process, I suspect you would see the same behavior. We are currently seeing the same thing when doing a straight in-process to out-of-process migration in a .NET 6 function application.
> 
> Thanks, I'll try that, guess still means we can't update to .NET 8 though, as .NET 8 doesn't support in-process to my knowledge.

You are correct.  .NET 8 is not supported for in-proc functions yet.  There is a tracking [issue](https://github.com/Azure/azure-functions-host/issues/9951) that indicates that support for Windows functions should be rolling out sometime this month (June).

**Comment by MarkMilfordMe** [comment]
> > > @MarkMilfordMe I don't think that this is a .NET 8 vs .NET 6 issue. It appears to be an in-process vs out-of-process function issue. Your .NET 6 demo project is in-process and your .NET 8 demo project is out-of-process. If you were to change your .NET 6 project to out-of-process, I suspect you would see the same behavior. We are currently seeing the same thing when doing a straight in-process to out-of-process migration in a .NET 6 function application.
> > 
> > 
> > Thanks, I'll try that, guess still means we can't update to .NET 8 though, as .NET 8 doesn't support in-process to my knowledge.
> 
> You are correct. .NET 8 is not supported for in-proc functions yet. There is a tracking [issue](https://github.com/Azure/azure-functions-host/issues/9951) that indicates that support for Windows functions should be rolling out sometime this month (June).

So just did a quick example (added to original repo) and yes it fails on .net 6 out of proc.  Will cross fingers that In-Proc .NET 8 rolls out in plenty of time for us to transistion before november when .net 6 is out of support!  Will subscribe to the issue you mentioned.  Thanks

**Comment by patrick-elmore** [comment]
We found a workaround for this that may work for others.  It involves a very minimal custom middleware component:

```
public class HttpContextAccessorMiddleware(IHttpContextAccessor httpContextAccessor) : IFunctionsWorkerMiddleware
{
    public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
    {
        httpContextAccessor.HttpContext = context.GetHttpContext();
        await next(context);
    }
}
```

Then wire it up in Program.cs

```
var host = new HostBuilder()
    .ConfigureFunctionsWebApplication(builder =>
    {
        builder.UseMiddleware<HttpContextAccessorMiddleware>();
    })
```

This should allow you to inject IHttpContextAccessor as you did prior to moving to isolated.  I have only confirmed this on .NET 6, but I assume it will work in 8 as well.

**Comment by MarkMilfordMe** [comment]
Yes this works with the middleware

**Comment by soonwong** [comment]
We are using the above middleware pattern in our .net8 out-of-process functions.  However it does not appear to be 100% parity with in-proc.

I'm doing the following to get access to Application Insights request telemetry so that we can insert custom properties.  `HttpContext` is available but this returns NULL still.
```
var telemetry = _httpContextAccessor.HttpContext?.Features.Get<RequestTelemetry>();
```

I'm aware that this isn't quite on point; just be aware that the above middleware suggestion may not solve everything.

**Comment by metjuperry** [comment]
> We found a workaround for this that may work for others. It involves a very minimal custom middleware component:
> 
> ```
> public class HttpContextAccessorMiddleware(IHttpContextAccessor httpContextAccessor) : IFunctionsWorkerMiddleware
> {
>     public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
>     {
>         httpContextAccessor.HttpContext = context.GetHttpContext();
>         await next(context);
>     }
> }
> ```
> 
> Then wire it up in Program.cs
> 
> ```
> var host = new HostBuilder()
>     .ConfigureFunctionsWebApplication(builder =>
>     {
>         builder.UseMiddleware<HttpContextAccessorMiddleware>();
>     })
> ```
> 
> This should allow you to inject IHttpContextAccessor as you did prior to moving to isolated. I have only confirmed this on .NET 6, but I assume it will work in 8 as well.

This workaround works, thank you very much, but it's a shame we still have to do this.

**Comment by joethepro22** [comment]
This seems still to be a problem in .NET 10. Is that correct?
